src/couch_replicator_connection.erl

```diff
@@ -0,0 +1,105 @@
+-module(couch_replicator_connection).
+
+-behavior(gen_server).
+
+-export([start_link/0]).
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2]).
+-export([code_change/3, terminate/2]).
+
+-export([acquire/1, relinquish/1]).
+
+-record(state, {}).
+
+-record(connection, {
+    host,
+    port,
+    worker,
+    owner
+}).
+
+-include_lib("ibrowse/include/ibrowse.hrl").
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #connection.worker}]),
+    process_flag(trap_exit, true),
+    {ok, #state{}}.
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
+        {ok, Worker} ->
+            link(Worker),
+            {ok, Worker};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+relinquish(Worker) ->
+    gen_server:cast(?MODULE, {relinquish, Worker}).
+
+
+handle_call({acquire, URL}, From, State) ->
+    {Pid, _Ref} = From,
+    WorkerPid = case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    {ok, WorkerPid0} = ibrowse:spawn_link_worker_process(URL),
+                    couch_log:error("Worker created! ~p", [WorkerPid0]),
+                    ets:insert(?MODULE, #connection{host=Host, port=Port, owner=Pid, worker=WorkerPid0}),
+                    WorkerPid0;
+                {[Worker], _Cont} ->
+                    couch_log:error("Worker reassigned! ~p -> ~p", [Worker#connection.worker, Pid]),
+                    ets:insert(?MODULE, Worker#connection{owner=Pid}),
+                    Worker#connection.worker
+            end
+    end,
+    monitor(process, Pid),
+    {reply, {ok, WorkerPid}, State}.
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    case ets:lookup(?MODULE, WorkerPid) of
+        [Worker] ->
+            couch_log:error("Worker relinquished! ~p", [WorkerPid]),
+            ets:insert(?MODULE, Worker#connection{owner=undefined});
+        [] ->
+            ok
+    end,
+    {noreply, State}.
+
+
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
+    lists:foreach(fun(Worker) ->
+        ets:insert(?MODULE, Worker#connection{owner=undefined})
+    end, Workers),
+    {noreply, State};
+
+% worker crashed
+handle_info({'EXIT', Pid, Reason}, State) ->
+    case ets:match_object(?MODULE, #connection{worker=Pid, _='_'}, 1) of
+        '$end_of_table' ->
+            ok;
```

Comment:
If the owner crashes (which is linked to the worker process) then the workers will be removed from the ets table before the worker crash messages are received.